lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,565 @@
+# trainer = Trainer(callbacks=[FinetuningScheduler()], resume_from_checkpoint="some/path/to/my_checkpoint.ckpt")
+# ```
+#
+# Training will resume at the depth/level of the provided checkpoint according the specified schedule. Schedules can be altered between training sessions but schedule compatibility is left to the user for maximal flexibility. If executing a user-defined schedule, typically the same schedule should be provided for the original and resumed training sessions.
+#
+# By default (``FinetuningScheduler.restore_best`` is ``True``), ``FinetuningScheduler`` will attempt to restore the best available checkpoint before finetuning depth transitions.
+#
+# ```python
+# trainer = Trainer(
+#     callbacks=[FinetuningScheduler(new_incarnation_mode=True)],
+#     resume_from_checkpoint="some/path/to/my_kth_best_checkpoint.ckpt",
+# )
+# ```
+#
+# To handle the edge case wherein one is resuming scheduled finetuning from a non-best checkpoint and the previous best checkpoints may not be accessible, setting ``FinetuningScheduler.new_incarnation_mode`` to
+# ``True`` as above will re-intialize the checkpoint state with a new best checkpoint at the resumption depth.
+
+# %% [markdown]
+# <div class="alert alert-warning">
+#
+# **Note:** Currently, _FinetuningScheduler_ only supports the following _TrainingTypePlugins_:
+#
+# - ``DDPPlugin``
+# - ``DDPShardedPlugin``
+# - ``DDPSpawnPlugin``
+# - ``DDPSpawnShardedPlugin``
+# - ``DataParallelPlugin``
+# - ``SingleDevicePlugin``
+#
+# </div>
+
+# %% [markdown]
+# <div id="superglue"></div>
+#
+# ## Scheduled Finetuning For SuperGLUE
+#
+# The following example demonstrates the use of ``FinetuningScheduler`` to finetune a small foundational model on the [RTE](https://huggingface.co/datasets/viewer/?dataset=super_glue&config=rte) task of [SuperGLUE](https://super.gluebenchmark.com/). Iterative early-stopping will be applied according to a user-specified schedule.
+#
+# ``FinetuningScheduler`` can be used to achieve non-trivial model performance improvements in both implicit and explicit scheduling contexts at an also non-trivial computational cost.
+#
+
+# %%
+import logging
+import os
+import warnings
+from datetime import datetime
+from importlib import import_module
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import datasets
+import pytorch_lightning as pl
+import torch
+from pytorch_lightning.loggers.tensorboard import TensorBoardLogger
+from pytorch_lightning.utilities import rank_zero_warn
+from pytorch_lightning.utilities.cli import CALLBACK_REGISTRY, _Registry
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from torch.utils.data import DataLoader
+from transformers import AutoConfig, AutoModelForSequenceClassification, AutoTokenizer
+
+# %%
+# a couple helper functions to prepare code to work with the forthcoming hub and user module registry
+MOCK_HUB_REGISTRY = _Registry()
```

Comment:
Let's drop all references to the hub, this isn't public knowledge yet :)

Reply:
What hub? :wink: Renamed to a generic mock registry.